### ISSUE TITLE
GRPC bidirectional streaming:  small messages (< 4kB) are buffered/delayed

### ISSUE THREAD

**Issue by ondrej-smola** [containous/traefik#4007]
### Do you want to request a *feature* or report a *bug*?

Bug

### What did you do?

Using Traefik (as a replacement for Envoy) for grpc bidirectional streaming API when there are client and server sending very small protobuf messages (e.g. few bytes) and "normal size" messages ( > 4kB). 

### What did you expect to see?

Both client and server should receive both message sizes without delay.

### What did you see instead?

Larger messages are received immediately, small messages got buffered for approx. 100ms and then send. 

### Output of `traefik version`: (_What version of Traefik are you using?_)

```
Version:      v1.7.2
Codename:     maroilles
Go version:   go1.11.1
Built:        2018-10-04_01:44:36PM
OS/Arch:      linux/amd64
```

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

```
Docker 18.06, Ubuntu 18.04
```

We have the same API implemented using Websocket (no problems here both with Traefik and Envoy) and also have the same API running behind Envoy without problems. Does not matter if using h2c or https with self signed cert + insecureSkipVerify. From what I tested it works as long as the message has more than 4096 bytes. Any smaller message (3kB ...) gets buffered and send with delay (consistently around 100ms - avg. RTT is 20ms in my setup). Ignoring Traefik and doing direct communication (grpc client - server) does not have this issue.

**Comment by Juliens** [comment]
Thanks for your interest in the project!

Unfortunately, I was unable to reproduce. Do you think you can share a server and a client to reproduce this behavior?

**Comment by ondrej-smola** [comment]
I will try to create a minimal solution for reproducing this issue (cannot share our server as it is part of our internal codebase).

**Comment by ondrej-smola** [comment]
https://github.com/ondrej-smola/traefik-grpc-debug

1. Start the traefik (1.7.2) in the project root
2. Start dummy GRPC server (recv, **wait 50ms** and reply with the same message)
```
./traefik-grpc-debug.exe -server
```
2. Start dummy GRPC client (sends small 64 byte and large 8096 byte message)
```
./traefik-grpc-debug.exe
```

My example **client** side log

```
ts=2018-10-15T12:01:49.7606676Z msg_out=42 len=64  (client sends)
ts=2018-10-15T12:01:49.909901Z msg_in=42 len=64  (client receives)
ts=2018-10-15T12:01:49.909901Z msg_out=43 len=8096
ts=2018-10-15T12:01:49.9610026Z msg_in=43 len=8096
ts=2018-10-15T12:01:49.9610026Z msg_out=44 len=64
ts=2018-10-15T12:01:50.1102672Z msg_in=44 len=64
ts=2018-10-15T12:01:50.1102672Z msg_out=45 len=8096
ts=2018-10-15T12:01:50.1613613Z msg_in=45 len=8096
ts=2018-10-15T12:01:50.1613613Z msg_out=46 len=64
ts=2018-10-15T12:01:50.3097845Z msg_in=46 len=64
ts=2018-10-15T12:01:50.3097845Z msg_out=47 len=8096
ts=2018-10-15T12:01:50.3598661Z msg_in=47 len=8096
ts=2018-10-15T12:01:50.3598661Z msg_out=48 len=64
ts=2018-10-15T12:01:50.5110815Z msg_in=48 len=64
ts=2018-10-15T12:01:50.5110815Z msg_out=49 len=8096
ts=2018-10-15T12:01:50.5621725Z msg_in=49 len=8096
```
Notice the difference in response time (e.g. 50ms (just server side sleep) vs 140ms delay)

**Comment by traefiker** [comment]
Closed by #4112.
